src/mongoc/mongoc-server-description.c

```diff
@@ -0,0 +1,483 @@
+ */
+
+mongoc_host_list_t *
+mongoc_server_description_host (mongoc_server_description_t *description)
+{
+   return &description->host;
+}
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * mongoc_server_description_set_state --
+ *
+ *       Change the state of this server.
+ *
+ * Returns:
+ *       true, false
+ *
+ * Side effects:
+ *       None
+ *
+ *--------------------------------------------------------------------------
+ */
+void
+mongoc_server_description_set_state (mongoc_server_description_t *description,
+                                     mongoc_server_description_type_t type)
+{
+   description->type = type;
+   // TODO SDAM unblock waiters? Can waiters wait on particular server?
+}
+
+
+/*
+ *-------------------------------------------------------------------------
+ *
+ * mongoc_server_description_update_rtt --
+ *
+ *       Update this server's rtt calculation using an exponentially-
+ *       weighted moving average formula.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       Changes this server description's rtt.
+ *
+ *-------------------------------------------------------------------------
+ */
+void
+mongoc_server_description_update_rtt (mongoc_server_description_t *server,
+                                      int64_t                      new_time)
+{
+   // TODO SS implement
+   return;
+}
+
+/*
+ *-------------------------------------------------------------------------
+ *
+ * Called during SDAM, from topology description's ismaster handler.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+void
+mongoc_server_description_handle_ismaster (
+   mongoc_server_description_t   *sd,
+   const bson_t                  *ismaster_response,
+   int64_t                        rtt_msec,
+   bson_error_t                  *error)
+{
+   bson_iter_t iter;
+   bool is_master = false;
+   bool is_shard = false;
+   bool is_secondary = false;
+   bool is_arbiter = false;
+   /*
+   bool is_passive = false;
+   bool is_hidden = false;
+   */
+   bool is_replicaset = false;
+   const uint8_t *bytes;
+   uint32_t len;
+
+   bson_return_if_fail (sd);
+   bson_return_if_fail (ismaster_response);
+
+   bson_destroy (&sd->last_is_master);
+   bson_copy_to (ismaster_response, &sd->last_is_master);
+
+   bson_iter_init (&iter, &sd->last_is_master);
+
+   memset (&sd->set_name, 0, sizeof (*sd) - ((char*)&sd->set_name - (char*)sd));
+
+   while (bson_iter_next (&iter)) {
+      /* TODO: do we need to handle maxBsonObjSize */
+      if (strcmp ("ismaster", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_BOOL (&iter)) goto ERROR;
+         is_master = bson_iter_bool (&iter);
+      } else if (strcmp ("maxMessageSizeBytes", bson_iter_key (&iter)) == 0) {
+         if (! BSON_ITER_HOLDS_INT32 (&iter)) goto ERROR;
+         sd->max_write_batch_size = bson_iter_int32 (&iter);
```

Comment:
The ismaster response's "maxWriteBatchSize" is the *number* of documents in a batch, not the number of bytes. It's distinct from maxMessageSizeBytes.